### PR TITLE
Little rework of ClosedScioContext

### LINKED ISSUE #2030: Remove isBlocking option
`isBlocking` option under `ScioOptions` should be removed in favor of explicitly adding `waitUntilDone()`. This option also doesn't work well with the new semantic that `ClosedScioContext` introduces.

As deprecation policy, 2 options:
* the presence of this option in the args should throw an error before running the pipeline.
* rename `.close()` to `.run()`

leaning towards point 2

### PR DESCRIPTION
**Pull request by regadas**
Fixes #2030

### CODE REVIEW

File: scio-core/src/main/scala/com/spotify/scio/ScioContext.scala [hunk 1]
```diff
@@ -550,8 +611,7 @@ class ScioContext private[scio] (val options: PipelineOptions, private var artif
     options.setJobName(name)
   }
 
-  /** Close the context. No operation can be performed once the context is closed. */
-  def close(): ClosedScioContext = requireNotClosed {
+  private[scio] def prepare(): ScioContext = requireNotClosed {
```
Comment (ClaireMcGinty) [inline scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: any reason this logic can't go in `ScioRunner` directly?

Comment (regadas) [reply scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: reason is to not leak `ScioContext` internals

File: scio-core/src/main/scala/com/spotify/scio/ScioContext.scala [hunk 2]
```diff
@@ -456,7 +517,7 @@ class ScioContext private[scio] (val options: PipelineOptions, private var artif
     }
 
   /** Amount of time to block job for. */
-  private[scio] val awaitDuration: Duration = {
+  val awaitDuration: Duration = {
```
[the remark this answers is not in the document]
Comment (regadas) [reply scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: it's not used anywhere but we were offering `getAwaitDuration` before ... so simply offering a way to get it

File: scio-core/src/main/scala/com/spotify/scio/ScioContext.scala [hunk 3]
```diff
@@ -233,6 +233,137 @@ object ContextAndArgs {
   }
 }
 
+trait ScioExecutor {
+  def run(sc: ScioContext): ScioExecutionContext
+}
+
+sealed trait ScioRunner extends ScioExecutor
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: What's the reason for having `ScioRunner` and `ScioExecutor`? They can be the same trait/object?

Comment (regadas) [reply scio-core/src/main/scala/com/spotify/scio/ScioContext.scala]: the idea is to allow the following:

```scala
val sc: ScioContext = ???
val testExecutor: ScioExecutor = ???
val runner = ScioRunner(testExecutor)

// case 1
sc.run(runner)

// case 2
runner.run(sc)
```

in either case inner context logic before `run` is preserved. User is only allowed to override the `ScioExecutor`

**Pull request merged**